src/odemis/driver/smaract.py

```diff
@@ -1278,7 +1282,11 @@ def __init__(self, name, role, locator, axes, ref_on_init=False, linear_speed=0.
 
     def terminate(self):
         # should be safe to close the device multiple times if terminate is called more than once.
-        self.core.SA_MC_Close(self._id)
+        if not self._shutdown_flag:
+            self.stop()
+            self._executor.shutdown(cancel_futures=True)
+            self.core.SA_MC_Close(self._id)
+            self._shutdown_flag = True
```

Comment:
I think you should be able to to away with the `_shutdown_flag`, by using `_executor is not None`. After shutting down the executor, set it to `._executor = None`.
(Same thing for both classes).